fulltext/controllers.py

```diff
@@ -1,105 +1,239 @@
+    # doing anything else.
+    if authorizer and not authorizer(identifier, product.owner):
+        raise NotFound('No such extraction')
+
+    if product.content is None \
+            and product.status is Extraction.Status.IN_PROGRESS:
+        target = url_for('fulltext.task_status', identifier=identifier,
+                         id_type=id_type)
+        return TASK_IN_PROGRESS, status.SEE_OTHER, {'Location': target}
+    return product.to_dict(), status.OK, {}
+
+
+def start_extraction(id_type: str, identifier: str, token: str,
+                     force: bool = False,
+                     authorizer: Optional[Authorizer] = None) -> Response:
+    """Handle a request to force text extraction."""
+    if id_type not in SupportedBuckets:
+        raise NotFound('Unsupported identifier')
+
+    canonical = pdf.CanonicalPDF.current_session()
+    storage = store.Storage.current_session()
+    compilations = compiler.Compiler.current_session()
+
+    if not force:
+        # If an extraction product or task already exists for this paper,
+        # redirect. Don't do the same work twice for a given version of the
+        # extractor.
+        logger.debug('Check for an existing product or task')
+        product: Optional[Extraction] = None
+        try:
+            product = storage.retrieve(identifier, bucket=id_type,
+                                       meta_only=True)
+        except IOError as e:
+            raise InternalServerError('Could not connect to backend') from e
+        except store.DoesNotExist:
+            pass
+
+        if product is not None:
+            # Redirect to either the task status endpoint or the finished
+            # extraction product.
+            return _redirect(product, authorizer)
+
+    logger.debug('No existing task nor extraction for %s', identifier)
+    # Before creating an extraction task, check that the intended document
+    # even exists. This gives the client a clear failure now, rather than
+    # waiting until the async task fails. At the same time, we'll also grab
+    # the owner (if there is one) so that we can authorize the request.
+    owner: Optional[str] = None
+    if id_type == SupportedBuckets.ARXIV and not canonical.exists(identifier):
+        logger.debug('No PDF for this resource exists')
+        raise NotFound('No such document')
+    elif id_type == SupportedBuckets.SUBMISSION:
+        try:
+            owner = compilations.owner(identifier, token)
+            logger.debug('Got owner %s', owner)
+        except compiler.exceptions.NotFound as e:
+            logger.debug('Compiler returned 404 Not Found for %s', identifier)
+            raise NotFound('No such document') from e
+
+    # Make sure that the client is authorized to work with this resource.
+    # The route may have passed in an authorizer function that works with
+    # the auth API to authorize the request.
+    if authorizer is not None and not authorizer(identifier, owner):
```

Comment:
It seems like mistakenly calling this function without a authorizer will skip this authorization check.

Maybe put this check in ```elif id_type == SupportedBuckets.SUBMISSION:```

Reply:
Fixed in bcbaf1b ; that block now looks like:

```python
def start_extraction(...) -> Response:
    ...
    # Before creating an extraction task, check that the intended document
    # even exists. This gives the client a clear failure now, rather than
    # waiting until the async task fails. At the same time, we'll also grab
    # the owner (if there is one) so that we can authorize the request.
    owner: Optional[str] = None
    if id_type == SupportedBuckets.ARXIV and not canonical.exists(identifier):
        logger.debug('No PDF for this resource exists')
        raise NotFound('No such document')

        # Make sure that the client is authorized to work with this resource.
        # The route may have passed in an authorizer function that works with
        # the auth API to authorize the request.
        if authorizer is not None and not authorizer(identifier, owner):
            logger.debug('Client is not authorized to work with this resource')
            # Pretend that the resource does not even exist.
            raise NotFound('No such document')
    elif id_type == SupportedBuckets.SUBMISSION:
        try:
            owner = compilations.owner(identifier, token)
            logger.debug('Got owner %s', owner)
        except compiler.exceptions.NotFound as e:
            logger.debug('Compiler returned 404 Not Found for %s', identifier)
            raise NotFound('No such document') from e

        # Make sure that the client is authorized to work with this resource.
        # For submissions, an authorizer function must be provided, so we
        # should deny access if one is mistakenly not provided.
        if authorizer is None or not authorizer(identifier, owner):
            logger.debug('Client is not authorized to work with this resource')
            # Pretend that the resource does not even exist.
            raise NotFound('No such document')
    else:
        raise NotFound('No such document')

    if not force:
        ...
```

Reply:
Oh whoops that has an error though; fixing now

Reply:
```python
    if id_type == SupportedBuckets.ARXIV:
        if not canonical.exists(identifier):
            logger.debug('No PDF for this resource exists')
            raise NotFound('No such document')

        # Make sure that the client is authorized to work with this resource.
        # The route may have passed in an authorizer function that works with
        # the auth API to authorize the request.
        if authorizer is not None and not authorizer(identifier, owner):
```